src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserRepository.cs

```diff
@@ -24,4 +52,180 @@ public async Task<User> FindDeletedUserPlaceholder(CancellationToken cancellatio
             cancellationToken
         );
     }
+
+    public async Task<User> CreateOrUpdate(
+        string email,
+        GlobalRoles.Role role,
+        Guid createdById,
+        DateTimeOffset? createdDate = null,
+        CancellationToken cancellationToken = default
+    ) =>
+        await CreateOrUpdate(
+            email: email,
+            roleId: role.GetEnumValue(),
+            createdById: createdById,
+            createdDate: createdDate,
+            cancellationToken: cancellationToken
+        );
+
+    public async Task<User> CreateOrUpdate(
+        string email,
+        string roleId,
+        Guid createdById,
+        DateTimeOffset? createdDate = null,
+        CancellationToken cancellationToken = default
+    )
+    {
+        if (createdDate > DateTimeOffset.UtcNow)
+        {
+            throw new ArgumentException($"{nameof(User)} created date cannot be a future date.");
+        }
+
+        var existingUser = await contentDbContext.Users.SingleOrDefaultAsync(i =>
+            i.Email.ToLower().Equals(email.ToLower())
+        );
+
+        return existingUser is null
+            ? await CreateNewUser(
+                email: email,
+                roleId: roleId,
+                createdById: createdById,
+                createdDate: createdDate,
+                cancellationToken: cancellationToken
+            )
+            : await UpdateExistingUser(
+                existingUser: existingUser,
+                roleId: roleId,
+                createdById: createdById,
+                createdDate: createdDate,
+                cancellationToken: cancellationToken
+            );
+    }
+
+    public async Task SoftDeleteUser(User activeUser, Guid deletedById, CancellationToken cancellationToken = default)
+    {
+        contentDbContext.Attach(activeUser);
+
+        activeUser.Active = false;
+        activeUser.SoftDeleted = DateTime.UtcNow;
+        activeUser.DeletedById = deletedById;
+
+        await contentDbContext.SaveChangesAsync(cancellationToken);
+    }
```

Comment:
Would it be simpler and more flexible for this to have a `Guid userId` parameter rather than `User activeUser`?

It would avoid `contentDbContext.Attach(activeUser);` and it might make unit testing easier too.